### PR TITLE
gcc: add common/dependencies.nix

### PR DESCRIPTION
**Pull request by Artturin**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/gcc/9/default.nix [hunk 1]
```diff
@@ -158,34 +158,31 @@ stdenv.mkDerivation ({
...
-    ++ optional targetPlatform.isLinux patchelf;
-
-  buildInputs = [
-    gmp mpfr libmpc
-    targetPackages.stdenv.cc.bintools # For linking code at run-time
-  ] ++ (optional (isl != null) isl)
-    ++ (optional (zlib != null) zlib)
+  inherit (import ../common/dependencies.nix {
+    inherit
+      lib
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gcc/9/default.nix:163]: Naive question: can we also arredicate this inherit list?

File: pkgs/development/compilers/gcc/common/dependencies.nix [hunk 2]
```diff
@@ -0,0 +1,92 @@
+{ lib
+, stdenv
+, version
+, buildPackages
+, targetPackages
+, texinfo
+, which
+, gettext
+, pkg-config ? null
+, gnused
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gcc/common/dependencies.nix:10]: I wouldn't mind if we remove as a next step some of those ``? null``s.

File: pkgs/development/compilers/gcc/9/default.nix (hunk 1 above)
Comment (Artturin) [reply pkgs/development/compilers/gcc/9/default.nix:163]: i tried passing `@args` but https://nixos.org/manual/nix/stable/language/constructs.html?highlight=%40args#functions

image

File: pkgs/development/compilers/gcc/common/dependencies.nix (hunk 2 above)
Comment (Artturin) [reply pkgs/development/compilers/gcc/common/dependencies.nix:10]: i think nulls are being used because of bootstrapping

**Comment by mweinelt** [comment]
A single sentence rationale would be appreciated.

**Comment by Artturin** [comment]
> A single sentence rationale would be appreciated. You seem to be aggregating dependencies between gcc versions?

added some commits msgs

> Why are the commits grouped like they are?

i thought it would help reviewers to show what i had to move for each version

**Review by trofi — APPROVED** [review]

**Review by Ericson2314 — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by amjoseph-nixpkgs** [comment]
Thank you so much for this.  Massive improvement.

File: pkgs/development/compilers/gcc/9/default.nix (hunk 1 above)
Comment (amjoseph-nixpkgs) [reply pkgs/development/compilers/gcc/9/default.nix:163]: I think the right way to do this is `lib.newScope`.

**Comment by yu-re-ka** [comment]
pkgsMusl.gccgo12 is now missing its libucontext buildInput.
I'm not sure what is the best way to proceed, since the `nix eval` command used to generate the list in pkgs/development/compilers/gcc/12/default.nix:140 will not print this dependency because it is conditional.
So if we add libucontext to the list now, how do we make sure it does not go missing the next time the list is regenerated?

**Comment by Artturin** [comment]
> pkgsMusl.gccgo12 is now missing its libucontext buildInput. I'm not sure what is the best way to proceed, since the `nix eval` command used to generate the list in pkgs/development/compilers/gcc/12/default.nix:140 will not print this dependency because it is conditional. So if we add libucontext to the list now, how do we make sure it does not go missing the next time the list is regenerated?

You meant to comment in https://github.com/NixOS/nixpkgs/pull/215751

I think I just copied the wrong jq list to gcc12 because 
`nix eval --json --impure --expr "with import ./. {}; lib.attrNames (lib.functionArgs gcc12.cc.override)" | jq '.[]' --raw-output` has libucontext

**Pull request merged**